src/backend/columnar/columnar_metadata.c

```diff
@@ -1295,10 +1299,59 @@ ColumnarStorageUpdateIfNeeded(Relation rel, bool isUpgrade)
 	GetHighestUsedAddressAndId(storageId, &highestOffset, &highestId);
 
 	uint64 reservedStripeId = highestId + 1;
-
-	/* XXX: should be set properly */
-	uint64 reservedRowNumber = 0;
 	uint64 reservedOffset = highestOffset + 1;
+
+	uint64 highestFirstRowNumber = COLUMNAR_INVALID_ROW_NUMBER;
+	uint64 stripeRowCount = -1;
+	GetHighestUsedFirstRowNumberAndRowCount(storageId, &highestFirstRowNumber,
+											&stripeRowCount);
+
+	uint64 reservedRowNumber = COLUMNAR_INVALID_ROW_NUMBER;
+	if (highestFirstRowNumber == COLUMNAR_INVALID_ROW_NUMBER)
+	{
+		/* relation has no stripes, use COLUMNAR_FIRST_ROW_NUMBER */
+		reservedRowNumber = COLUMNAR_FIRST_ROW_NUMBER;
+	}
+	else
+	{
+		/* calculate reservedRowNumber for the next stripe */
+		reservedRowNumber = highestFirstRowNumber + stripeRowCount;
+	}
```

Comment:
Wouldn't it be simpler to just have GetHighestUsed...() return the highest used row number, a single value, and you wouldn't have to check for invalid values here or add the two numbers together?

Reply:
done